Update distributed programming 07 06 20 (#229)
Agoric/documentation: main/distributed-programming.md

# Agoric JavaScript Programming Extensions

> Agoric's platform lets you write secure smart contracts in JavaScript. The platform itself is mainly written in JavaScript. However, we've made several Agoric-specific additions to general JavaScript programming that you should know about and understand before programming on the platform. Some are *concepts*, others are *Agoric library additions*, and some are at the *syntax level*.

Reviewer: I'd add that all changes at the language level are in process to becoming official standards.

- **[Vats](#vats)**: Objects and functions In the same JavaScript vat can
  communicate synchronously. Communication with objects outside the
  vat can only be done asynchronously. 

- **Secure EcmaScript (SES)**: SES provides a secure platform for
  executing programs. With SES, you can run code you don't completely trust,
  without being vulnerable to bugs or bad intentions. It's a
  standards-track extension to the JavaScript standard. 

- **Realms and compartments**: JavaScript code runs in the context of
  a *realm*, made up of the set of primordials (the standard
  library's objects and functions) and a global object. In a web
  browser, an iframe is a realm. In Node.js, a Node process is a
  realm. A *compartment* is a separate execution environment within a realm.

- **`harden()`**: A hardened object’s properties cannot be changed, so the only way to interact with a hardened object is through its methods. `harden()`is similar to `Object.freeze()` but
more powerful. 

- **Remote object communication using `E`**: `E` is a local "bridge" function that lets you invoke methods on remote objects, whether in another vat, machine, or blockchain (for example).

> It takes a local representative (a *proxy*) for a remote object as an argument and sends messages to it using normal message-sending syntax. The local proxy forwards all messages to the remote object to deal with. If you are familiar with the `~` operator, `E` is a workaround for the JavaScript standard not yet supporting `~`.

Reviewer: It's not the "\~" operator, it's "\~.", pronounced til-dot, by decree of a twitter poll.

Reviewer 2: The JS standards proposal document uses the more grownup-sounding "wavy dot": https://github.com/tc39/proposal-wavy-dot . But all its friends use "tildot".

- **Notifiers:** Our Promise-based Notifier notifies Dapps and other tools
about changes to their subscribed to contracts or offers' state.

## Vats

A vat is a *unit of isolation*. To paraphrase the Las Vegas advertising slogan, what happens in the vat stays in the vat. Objects and functions in a JavaScript vat can communicate synchronously with one another. Vats and their contents can communicate with other vats and their objects and functions, but have to
[manage asynchronous messages and responses](#Communicating-with-remote-objects-via-E).

There are no tools for telling what vat something is in, or if two things are in the same or different vats. In general, you/your code should know if things are local (in the same vat) because you created them or they were passed to you by something guaranteeing that’s the case. Other objects you should treat as if they might be distant (in different vats). 

Vats need to run on some platform. Both a single physical machine and a blockchain (which might itself be running on a set of  collaborating machines) are possible platforms. Either type of platform can host one or more vats.

> Since a vat runs in a single *event loop*, each incoming request has to finish before the next one starts. If there's remaining work, you schedule it to happen later when a Promise resolves.

Reviewer: Yes, what Kate said.

The Agoric process starts several vats. Each vat hosts a service (e.g. Registry, Zoe,
etc.). As of April, 2020, all contracts run in the Zoe vat. Eventually this will change to each contract having a dedicated vat.

## Secure EcmaScript (SES)

SES is a standards-track extension of the JavaScript standard 
[EcmaScript](https://www.ecma-international.org/publications/standards/Ecma-262.htm),

> It provides a secure platform for executing programs, letting you run code you don't completely trust. Our code, as operator of a chain machine, can host untrusted developer apps’ code without being vulnerable to possible bugs or bad intentions.  

Reviewer 3: ```suggestion
It provides a secure platform for executing programs, letting you run code you don't completely trust. SES-based programs, whether you are an end-user or a server operator, can host untrusted developer apps’ code without being vulnerable to possible bugs or bad intentions.  
```

For historical reasons, the ECMAScript specification requires the *primordials*
to be mutable (`Array.prototype.push = yourFunction` is valid
ECMAScript but not recommended). SES makes it possible to turn the
current realm into an **immutable realm**, that is a realm within
which the primordials are deeply frozen. It also allows programs to
create **Compartments**. 

Compartments are "mini-realms". They have their own dedicated global object and environment, but they inherit the primordials from their parent realm.

Agoric deploy scripts and smart contract code run in an immutable
realm with compartments providing just enough authority to create
useful and secure contracts (but not enough authority to do anything
unintended or harmful to the participants of the smart contract). 
This means you cannot send email, write to disk, or visit arbitrary web

> pages, etc. from code inside a contract.

Reviewer: code cannot do ... unless it was started by a process that had authority to do so, and explicitly passed that authority in.

## `harden()`

`harden()` ensures that external callers of a hardened object can only
interact with it through functions in the object’s API.
 `harden()` is an enhanced transitive version of 
[`Object.freeze`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze),
which only locks up an object's own properties.

All objects that will be transferred across a trust boundary must have
their API surface frozen (usually by calling `harden()`). This ensures
other objects can only interact with them through their defined
method interface. CapTP, our communications layer for passing
references to distributed objects, enforces this at vat boundaries.

The general rule is that if you make a new object and give it to someone else (and don't immediately forget it yourself),

> you should give them `harden(obj)` instead the raw object.

Reviewer: "instead the" => "instead of the"

This prevents someone from adding/deleting the properties or prototypes of that object.

> It doesn't make them immutable (`harden(new Map()) ` still behaves like a normal mutable `Map`), but it means

Reviewer 4: Should we maybe further define what we mean by immutable? I think just another sentence would add a lot. For example, if I were reading this, I would say "huh, preventing add/deleting the properties or prototype and not being able to change the values sounds like immutability, so what's the difference?"

Reviewer 2: maybe "Being hardened doesn't preclude an object from having access to mutable state (map example)", and make sure we point out the *surprising* mutability that is prevented with `harden`, as opposed to the "that's the whole point of a Map" -kind of mutability.

their methods stay the same and can't be surprisingly changed by someone else

Defined objects (`mint`, `issuer`, `zcf,` etc.) shouldn't need hardening as their constructors should do that work. It's mainly records, callbacks, and ephemeral objects that need hardening.

> You can send a message to a hardened object. If it's a record, you can extract values from it.

Reviewer 4: What does `extract values` mean? Don't we just mean accessing the properties and getting the values?

You have to harden a class before you harden any of its instances; i.e. it takes two separate steps
to harden both a class and its instances. `harden()` does transitive freezing by following the object’s
own properties (as opposed to properties it inherited), and the objects whose own properties refer to them,

> and so forth. Harden a base class before hardening classes that inherit from it.

Reviewer 4: I think we talk about classes and then how the freezing works, and then classes again. Can we rearrange?

`harden()` is provided by the
[`@agoric/harden` package](https://www.npmjs.com/package/@agoric/harden). 
After you've [installed](https://docs.npmjs.com/cli/install) the

> [`@agoric/harden` package](https://www.npmjs.com/package/@agoric/harden),

Reviewer 4: @warner should be able to comment on this, but I believe harden is provided by SES now

Reviewer 2: Yes, any code that will run inside a vat or a contract can use `harden` as a global, without importing anything.

And you can probably add `/* global harden */` to the top of your file to make your text editor stop complaining about `harden` not being defined or imported anywhere.

o.a  = 37; // throws a TypeError because o is now hardened
```

## Communicating with remote objects using `E`

On the Agoric platform, objects may be running in distinct vats, on a remote
machine, or even on a blockchain. When you send messages to non-local
objects in different vats, the response isn't received immediately and
can't be acted upon locally until it arrives. 

To keep from blocking local code until the response arrives, we
return a `Promise` for the result. You can send more messages to a result's
`Promise`. If and when the `Promise` resolves to a remote object, the messages

> are forwarded to the object's location, and their results arel

Reviewer 4: ```suggestion
are forwarded to the object's location, and their results are
```

eventually returned and processed locally. 

JavaScript natively
[supports Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises).
Agoric's `HandledPromises` are compatible and interoperable with
standard `Promise`s. Standard interaction with a `Promise` or
`HandledPromise` is to do further processing either in a `.then()`
clause or after using `await` to get the result. 

You can send messages to a `Promise`'s eventual result, or to a
presence (a proxy for a remote object), using `E()` notation. For
example, `E(remoteServiceP).startup(params)`. The result of a
send using`E` is always a `Promise`, so the normal thing to do with
the result (as with any object) is either pass it as a parameter or
invoke a function to be performed once the `Promise` is fulfilled.

`E(remoteServiceP).startup(params).then(result => useTheService(result));`

Deploy scripts and Zoe smart contracts often access services running in a
different vat. For instance, a deploy script may want to install a contract in a
Zoe instance running in a blockchain. But the deploy script

> cannot call `zoe.install(code, moduleFormat)`, because it does not have local

Reviewer 4: ```suggestion
cannot call `zoe.install(bundle)`, because it does not have local
```

access to the `zoe` object in a different vat. However, the deploy
script is given access to a `zoe` *presence*. To call methods on the
actual Zoe object, the deploy code can do:

```js

> const installationHandle = await E(zoe).install(source, moduleFormat);

Reviewer 4: ```suggestion
const installationHandle = await E(zoe).install(bundle);
```

```

The `E()` function is a local "bridge" that lets you invoke methods on
remote objects. The local version of a remote object is called a

> **presence**. `E()` takes a presence as an argument and creates an

Reviewer: The object created by E() is actually a forwarder that doesn't know what methods the remote object has. This is useful to know for debugging purposes. If you mis-spell or mis-capitalize the method name, the local environment can't tell; you'll only find out at run-time when the remote object complains that it doesn't know that method.

> object with the corresponding methods.

Reviewer 4: Huh, is this really how it works?

Reviewer 2: Nope. Chris's description is correct, `E()` doesn't know what the remote methods are, it just sends a message with whatever method name you used, and the target can complain if it doesn't have a method by that name.

`E()` performs the communication asynchronously. Method calls can take
objects in the current vat or presences for objects in other vats as arguments.

> `E()` is frequently used in contract code to call

Reviewer 4: ```suggestion
`E()` is frequently used in code to call
```
It's mostly used outside of contract code

[Zoe Service API methods](https://agoric.com/documentation/zoe/api/zoe.html).

## Notifiers

Dapps and other tools may want to be notified about state changes of a Zoe
contract or offer. Agoric uses a Notifier based on Promises, allowing
many subscribers to receive notifications without the publisher 
having to track a subscription list.

Zoe supports the Notifier, which publishes updates to offer state
(reallocations and completions). Some contracts also use it, and

> can publish current prices or other contract-specific details

Reviewer 4: ```suggestion
can publish current prices or other contract-specific details.
```

### Getting notifications

Zoe has a public method `getOfferNotifier()`, and contracts will have
similar methods. This method provides a long-lived notifier object associated 
with a particular stream of updates.

```js
const offerNotifer = zoe.getOfferNotifier(offerHandle);
  const { value, updateHandle, done } = offerNotifier.getUpdateSince();
  if (done) {
   <drop offer from list>
  }

>   newValue = value;

Reviewer 4: what is this line doing?

  waitForNextUpdate(offerNotifier, updateHandle);
```

Note: There is both a `zoe.getOfferNotifier()` and a `zcf.getOfferNotifier()`. Use the `zcf.` version 
within contracts and the `zoe.` version in the REPL, deploy scripts, and similar outside of a contract cases. 

When called on a notifier object`notifier.getUpdateSince()` returns
the record `{ value, updateHandle, done }`. `value` represents 
the state of an offer or contract. If you get a notifier from Zoe, you have to 

> identify the offer. 

Reviewer: "If you get a notifier from Zoe, you have to identify the offer. "
should be either

"If you want a notifier from Zoe, you have to identify the offer. "
or 
"To get a notifier from Zoe, you have to identify the offer. "

- `value` is the current state, according to the source. 
- `done` is `false` until the stream of updates reaches a final state. Then 

> `value` never changes and , and `getUpdateSince()` always returns the

Reviewer: redundant "and , and "
